File: OpenLive-macOS-Objective-C/OpenLive/SettingsViewController.m
//
//  SettingViewController.m
//  OpenLive
//
//  Created by yangmoumou on 2018/2/8.
//  Copyright © 2018年 yangmoumou. All rights reserved.
//

#import "SettingsViewController.h"
#import "MediaInfo.h"
#import "MainViewController.h"

@interface SettingsViewController ()
@property (weak) IBOutlet NSPopUpButton *profilePopUpButton;
@property (nonatomic,strong) NSArray *profiles;
@end

@implementation SettingsViewController
- (NSArray *)profiles {
    if (!_profiles) {
        _profiles = @[@(AgoraVideoDimension160x120),
                      @(AgoraVideoDimension320x240),
                      @(AgoraVideoDimension640x360),
                      @(AgoraVideoDimension640x480),
                      @(AgoraVideoDimension1280x720)];
    }
    return _profiles;
}

- (void)setVideoProfile:(CGSize)videoProfile {
    _videoProfile = videoProfile;
    [self loadProfies];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self loadProfies];
}

- (void)loadProfies {
    for (NSNumber *profile in self.profiles) {
        [self.profilePopUpButton addItemWithTitle:[MediaInfo descriptionProfile:[profile sizeValue]]];
    }
    [self.profilePopUpButton selectItemWithTitle:[MediaInfo descriptionProfile:self.videoProfile]];
}

- (IBAction)doProfileChanged:(NSPopUpButton *)sender {
    self.videoProfile = [self.profiles[sender.indexOfSelectedItem] sizeValue];
}

- (IBAction)doConfirmClicked:(NSButton *)sender {
    if ([self.delegate respondsToSelector:@selector(settingsVC:didSelectProfile:)]) {
        [self.delegate settingsVC:self didSelectProfile:self.videoProfile];
    }
}
@end

File: OpenLive-macOS/OpenLive/SettingsViewController.swift
//
//  SettingsViewController.swift
//  OpenLive
//
//  Created by GongYuhua on 2/20/16.
//  Copyright © 2016 Agora. All rights reserved.
//

import Cocoa

protocol SettingsVCDelegate: class {
    func settingsVC(_ settingsVC: SettingsViewController, closeWithProfile videoProfile: CGSize)
}

class SettingsViewController: NSViewController {

    @IBOutlet weak var profilePopUpButton: NSPopUpButton!
    
    var videoProfile: CGSize!
    var delegate: SettingsVCDelegate?
    fileprivate let profiles: [CGSize] = CGSize.validProfileList()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        loadProfileItems()
    }
    
    @IBAction func doProfileChanged(_ sender: NSPopUpButton) {
        let profile = profiles[sender.indexOfSelectedItem]
        videoProfile = profile
    }
    
    @IBAction func doConfirmClicked(_ sender: NSButton) {
        delegate?.settingsVC(self, closeWithProfile: videoProfile)
    }
}

private extension SettingsViewController {
    func loadProfileItems() {
        profilePopUpButton.addItems(withTitles: profiles.map { (res) -> String in
            return res.description()
        })
        profilePopUpButton.selectItem(withTitle: videoProfile.description())
    }
}
